In [1]:
!pip install neuralforecast

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.6/170.6 kB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 776.9/776.9 kB 9.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.6/62.6 MB 12.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 413.4/413.4 kB 28.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 806.1/806.1 kB 51.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.7/101.7 kB 6.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 230.6/230.6 kB 11.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 7.6 MB/s eta 0:00:00


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math
import numpy as np
import torch
from typing import Optional, Union, Tuple


import torch.nn.functional as F
from torch.distributions import Distribution
from torch.distributions import Bernoulli, Normal, StudentT, Poisson, NegativeBinomial

from torch.distributions import constraints


from neuralforecast import NeuralForecast
from neuralforecast.models import Autoformer
from neuralforecast.tsdataset import TimeSeriesDataset
from neuralforecast.utils import AirPassengers, AirPassengersPanel, AirPassengersStatic
from neuralforecast.losses.numpy import rmse, mape
from pandas.tseries.holiday import USFederalHolidayCalendar
from pandas.tseries.offsets import CustomBusinessDay
from neuralforecast.losses.pytorch import MQLoss, DistributionLoss, MSE, MAE

In [3]:
from numpy.random import seed
from random import randrange

In [4]:
df = pd.read_excel("SP100_vol (clean).xlsx", index_col="Date",parse_dates=True)
columns=df.columns
df_static=pd.read_excel("Static_df.xlsx")
n_series=len(df.columns)
train = df.iloc[:int(len(df['NKE'])*0.7)-1]
validation_length=int(len(df['NKE'])*0.7)-int(len(df['NKE'])*0.5)
train=train.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

In [5]:
n_inputs = [5,10,15,21,42,84]
hidden_size = [60,90,120,150,180]
epochs=[50,100,150,250,350,450,550,650,750]
learning_rate=[0.0005,0.0001,0.00005,0.00001]
num_lr_decays=[5,3,2,1]
scaler_type=["robust","standard",'minmax']
losses=[MSE(),MAE(),MQLoss(level=[80, 90]),DistributionLoss(distribution='StudentT', level=[80, 90])]
n_head=[2,3,4,5,6]
decoder_input_size_multiplier=[0.3,0.4,0.5,0.6,0.7]
dropout=[0,0.2,0.3,0.4]
factor=[2,3,4]
conv_hidden_size=[12,22,32,42,52]
decoder_layers = [1,2,3]

# Hyperparameter Search

In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  if n_inputs[i]==5:
    a2=0
  if n_inputs[i]==15 or 10:
    a2=0
  else:
    a2=a2
  model = Autoformer(h=1, input_size=n_inputs[i],
             decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
             decoder_layers=decoder_layers[a5],
             MovingAvg_window=15, loss=losses[m],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | MovingAvg_window: {15}| conv_hidden_size:{conv_hidden_size[a3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 4703606


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 56674659


Epoch: 250 | Inputs: 15 | hidden_size: 120 | Loss: MSE() | Scaler Type: minmax | factor: 2 | decoder_layers: 1
Learning Rate: 1 | decoder_input_size_multiplier: 0.3 | Dropout: 0 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.0030055560779024553 | QLIKE: 0.03341386643676348 | MAE: 0.001967145900035732 | MAPE: 0.2015522773990321


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 47959064


Epoch: 550 | Inputs: 21 | hidden_size: 120 | Loss: MSE() | Scaler Type: minmax | factor: 2 | decoder_layers: 1
Learning Rate: 3 | decoder_input_size_multiplier: 0.4 | Dropout: 0 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:12
RMSE: 0.004768431771212712 | QLIKE: 0.03189205819059983 | MAE: 0.003312399276982608 | MAPE: 0.34170861575939687


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 23638022


Epoch: 50 | Inputs: 42 | hidden_size: 90 | Loss: DistributionLoss() | Scaler Type: standard | factor: 2 | decoder_layers: 2
Learning Rate: 1 | decoder_input_size_multiplier: 0.7 | Dropout: 0.4 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:32
RMSE: 0.0035252794438507973 | QLIKE: 0.04595828354857277 | MAE: 0.0022898552756400793 | MAPE: 0.21860404242036896


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 92660464


Epoch: 550 | Inputs: 10 | hidden_size: 180 | Loss: DistributionLoss() | Scaler Type: minmax | factor: 2 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.5 | Dropout: 0.4 | n_head: 3 | MovingAvg_window: 15| conv_hidden_size:42
RMSE: 0.0026701403110529504 | QLIKE: 0.025174291721614485 | MAE: 0.0016152744064378956 | MAPE: 0.15491464614494924


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 250 | Inputs: 5 | hidden_size: 60 | Loss: MQLoss() | Scaler Type: standard | factor: 2 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.3 | Dropout: 0.3 | n_head: 2 | MovingAvg_window: 15| conv_hidden_size:52
RMSE: 0.0028467213100262464 | QLIKE: 0.027333249667740724 | MAE: 0.0017984156403297639 | MAPE: 0.18038672966811692


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  if n_inputs[i]==5:
    a2=0
  if n_inputs[i]==15 or 10:
    a2=0
  else:
    a2=a2
  model = Autoformer(h=1, input_size=n_inputs[i],
             decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
             decoder_layers=decoder_layers[a5],
             MovingAvg_window=15, loss=losses[m],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | MovingAvg_window: {15}| conv_hidden_size:{conv_hidden_size[a3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 33573686


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 55992325


Epoch: 150 | Inputs: 15 | hidden_size: 60 | Loss: MAE() | Scaler Type: standard | factor: 2 | decoder_layers: 2
Learning Rate: 1 | decoder_input_size_multiplier: 0.4 | Dropout: 0.2 | n_head: 2 | MovingAvg_window: 15| conv_hidden_size:12
RMSE: 0.0029786337828895735 | QLIKE: 0.034649598221347236 | MAE: 0.0018842803924603109 | MAPE: 0.18217967401889504


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 115218202


Epoch: 150 | Inputs: 15 | hidden_size: 150 | Loss: MSE() | Scaler Type: standard | factor: 2 | decoder_layers: 3
Learning Rate: 2 | decoder_input_size_multiplier: 0.4 | Dropout: 0 | n_head: 2 | MovingAvg_window: 15| conv_hidden_size:42
RMSE: 0.002744938411099357 | QLIKE: 0.02725602567135747 | MAE: 0.0017529124542917992 | MAPE: 0.1775824416797229


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 66042157


Epoch: 650 | Inputs: 42 | hidden_size: 180 | Loss: MSE() | Scaler Type: standard | factor: 2 | decoder_layers: 1
Learning Rate: 1 | decoder_input_size_multiplier: 0.7 | Dropout: 0 | n_head: 3 | MovingAvg_window: 15| conv_hidden_size:32
RMSE: 0.0030285155405341 | QLIKE: 0.03267769083613879 | MAE: 0.0019476796837111696 | MAPE: 0.1929083872163408


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 85342717


Epoch: 150 | Inputs: 42 | hidden_size: 180 | Loss: MQLoss() | Scaler Type: robust | factor: 2 | decoder_layers: 1
Learning Rate: 1 | decoder_input_size_multiplier: 0.4 | Dropout: 0.3 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.0032994799632979667 | QLIKE: 0.038479486602402524 | MAE: 0.002132120386804505 | MAPE: 0.20866905821753012


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 350 | Inputs: 84 | hidden_size: 120 | Loss: MAE() | Scaler Type: robust | factor: 2 | decoder_layers: 1
Learning Rate: 2 | decoder_input_size_multiplier: 0.6 | Dropout: 0.2 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.0033277225505547064 | QLIKE: 0.03944440365864936 | MAE: 0.0021437464355175866 | MAPE: 0.20881335392104441


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  if i ==5:
    a2=2
  else:
    a2=a2
  model = Autoformer(h=1, input_size=n_inputs[i],
             decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
             decoder_layers=decoder_layers[a5],
             MovingAvg_window=15, loss=losses[m],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | MovingAvg_window: {15}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 47481525


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 66578269


Epoch: 150 | Inputs: 5 | hidden_size: 180 | Loss: MAE() | Scaler Type: robust | factor: 2 | decoder_layers: 1
Learning Rate: 2 | decoder_input_size_multiplier: 0.6 | Dropout: 0.2 | n_head: 4 | MovingAvg_window: 15
RMSE: 0.0026875316121796134 | QLIKE: 0.025207790827335264 | MAE: 0.0016256058653595106 | MAPE: 0.15748754654856273


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 40490282


Epoch: 150 | Inputs: 15 | hidden_size: 150 | Loss: MQLoss() | Scaler Type: minmax | factor: 2 | decoder_layers: 2
Learning Rate: 2 | decoder_input_size_multiplier: 0.7 | Dropout: 0.3 | n_head: 6 | MovingAvg_window: 15
RMSE: 0.003169022905515298 | QLIKE: 0.00032540520051144997 | MAE: 0.00206023699153959 | MAPE: 0.20570244803714777


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 122697361


Epoch: 100 | Inputs: 10 | hidden_size: 150 | Loss: MAE() | Scaler Type: minmax | factor: 3 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.7 | Dropout: 0.2 | n_head: 5 | MovingAvg_window: 15
RMSE: 0.0029181177431574523 | QLIKE: 0.032673853010548023 | MAE: 0.0018415839906147846 | MAPE: 0.18142411556087681


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 81713458


Epoch: 750 | Inputs: 15 | hidden_size: 150 | Loss: MAE() | Scaler Type: standard | factor: 4 | decoder_layers: 3
Learning Rate: 0 | decoder_input_size_multiplier: 0.6 | Dropout: 0.2 | n_head: 6 | MovingAvg_window: 15
RMSE: 0.0026666464455100025 | QLIKE: 0.025441424134900022 | MAE: 0.0015952964505629784 | MAPE: 0.1531693132362037


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 50 | Inputs: 15 | hidden_size: 60 | Loss: DistributionLoss() | Scaler Type: standard | factor: 3 | decoder_layers: 1
Learning Rate: 1 | decoder_input_size_multiplier: 0.4 | Dropout: 0.4 | n_head: 3 | MovingAvg_window: 15
RMSE: 0.003006851447221786 | QLIKE: 0.03312854816576531 | MAE: 0.001857494526031283 | MAPE: 0.17341513600432626


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  if n_inputs[i]==5:
    a2=0
  if n_inputs[i]==15:
    a2=0
  else:
    a2=a2
  model = Autoformer(h=1, input_size=n_inputs[i],
             decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
             decoder_layers=decoder_layers[a5],
             MovingAvg_window=15, loss=losses[m],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | MovingAvg_window: {15}| conv_hidden_size:{conv_hidden_size[a3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 63165403


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 73683851


Epoch: 450 | Inputs: 21 | hidden_size: 180 | Loss: MQLoss() | Scaler Type: minmax | factor: 3 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.3 | Dropout: 0.3 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:42
RMSE: 0.0032202876463568123 | QLIKE: 0.042185674423618966 | MAE: 0.0020963790703585105 | MAPE: 0.20784553997430763


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 107886063


Epoch: 100 | Inputs: 10 | hidden_size: 180 | Loss: MAE() | Scaler Type: standard | factor: 2 | decoder_layers: 3
Learning Rate: 3 | decoder_input_size_multiplier: 0.5 | Dropout: 0.2 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:52
RMSE: 0.0030208442087674577 | QLIKE: 0.032051868557748026 | MAE: 0.0019233098810201733 | MAPE: 0.19076165186718103


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 22184045


Epoch: 250 | Inputs: 10 | hidden_size: 180 | Loss: DistributionLoss() | Scaler Type: standard | factor: 3 | decoder_layers: 3
Learning Rate: 2 | decoder_input_size_multiplier: 0.4 | Dropout: 0.4 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:42
RMSE: 0.0027304285274264532 | QLIKE: 0.027029711611160002 | MAE: 0.001694763161408811 | MAPE: 0.1642102054621675


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 129141199


Epoch: 550 | Inputs: 21 | hidden_size: 180 | Loss: MSE() | Scaler Type: robust | factor: 2 | decoder_layers: 3
Learning Rate: 1 | decoder_input_size_multiplier: 0.6 | Dropout: 0 | n_head: 5 | MovingAvg_window: 15| conv_hidden_size:12
RMSE: 0.002698051713677184 | QLIKE: 0.025612513781248576 | MAE: 0.0016925917968366956 | MAPE: 0.17006410500086414


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 550 | Inputs: 42 | hidden_size: 180 | Loss: MAE() | Scaler Type: standard | factor: 4 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.6 | Dropout: 0.2 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.0031114167718332194 | QLIKE: 0.04324932144426737 | MAE: 0.0019893844050930497 | MAPE: 0.19083716206858486


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  if n_inputs[i]==5:
    a2=0
  if n_inputs[i]==15:
    a2=0
  else:
    a2=a2
  model = Autoformer(h=1, input_size=n_inputs[i],
             decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
             decoder_layers=decoder_layers[a5],
             MovingAvg_window=15, loss=losses[m],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | MovingAvg_window: {15}| conv_hidden_size:{conv_hidden_size[a3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 80132587


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 78847134


Epoch: 650 | Inputs: 15 | hidden_size: 120 | Loss: MSE() | Scaler Type: robust | factor: 2 | decoder_layers: 2
Learning Rate: 2 | decoder_input_size_multiplier: 0.5 | Dropout: 0 | n_head: 5 | MovingAvg_window: 15| conv_hidden_size:12
RMSE: 0.0026764254173243423 | QLIKE: 0.02527909165064844 | MAE: 0.0016866795972404983 | MAPE: 0.17000560940680517


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 289990


Epoch: 650 | Inputs: 10 | hidden_size: 180 | Loss: MSE() | Scaler Type: minmax | factor: 2 | decoder_layers: 2
Learning Rate: 1 | decoder_input_size_multiplier: 0.7 | Dropout: 0 | n_head: 5 | MovingAvg_window: 15| conv_hidden_size:12
RMSE: 0.0028117765155389668 | QLIKE: 0.027613402621164608 | MAE: 0.0017813890306880872 | MAPE: 0.17882844442196713


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 48186372


Epoch: 750 | Inputs: 5 | hidden_size: 120 | Loss: MAE() | Scaler Type: robust | factor: 2 | decoder_layers: 2
Learning Rate: 0 | decoder_input_size_multiplier: 0.4 | Dropout: 0.2 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:32
RMSE: 0.0026014803693844026 | QLIKE: 0.023928809728980475 | MAE: 0.0015433653548939167 | MAPE: 0.1489220579925518


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 106898867


Epoch: 650 | Inputs: 84 | hidden_size: 120 | Loss: MSE() | Scaler Type: minmax | factor: 2 | decoder_layers: 3
Learning Rate: 3 | decoder_input_size_multiplier: 0.7 | Dropout: 0 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.004136906767601027 | QLIKE: -0.0006733419923072544 | MAE: 0.0028297134642307747 | MAPE: 0.28584287775690786


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 100 | Inputs: 15 | hidden_size: 150 | Loss: MSE() | Scaler Type: standard | factor: 2 | decoder_layers: 3
Learning Rate: 0 | decoder_input_size_multiplier: 0.3 | Dropout: 0 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:12
RMSE: 0.0028266346445326254 | QLIKE: 0.0291242973615545 | MAE: 0.001816219356722326 | MAPE: 0.18342321703352316


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  if n_inputs[i]==5:
    a2=0
  if n_inputs[i]==15:
    a2=0
  else:
    a2=a2
  model = Autoformer(h=1, input_size=n_inputs[i],
             decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
             decoder_layers=decoder_layers[a5],
             MovingAvg_window=15, loss=losses[m],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | MovingAvg_window: {15}| conv_hidden_size:{conv_hidden_size[a3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 56597732


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 38167151


Epoch: 50 | Inputs: 42 | hidden_size: 120 | Loss: MAE() | Scaler Type: standard | factor: 4 | decoder_layers: 2
Learning Rate: 0 | decoder_input_size_multiplier: 0.6 | Dropout: 0.2 | n_head: 3 | MovingAvg_window: 15| conv_hidden_size:12
RMSE: 0.003203826523550131 | QLIKE: 0.03743666367728801 | MAE: 0.0020349764354259513 | MAPE: 0.1957652336509428


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 49733760


Epoch: 50 | Inputs: 42 | hidden_size: 90 | Loss: MQLoss() | Scaler Type: minmax | factor: 3 | decoder_layers: 3
Learning Rate: 2 | decoder_input_size_multiplier: 0.6 | Dropout: 0.3 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:32
RMSE: 0.006083463382355 | QLIKE: 0.09071201831769288 | MAE: 0.004294275418365356 | MAPE: 0.4662018404864138


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 89735082


Epoch: 150 | Inputs: 5 | hidden_size: 180 | Loss: MSE() | Scaler Type: standard | factor: 2 | decoder_layers: 3
Learning Rate: 3 | decoder_input_size_multiplier: 0.7 | Dropout: 0 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:52
RMSE: 0.00268097353115542 | QLIKE: 0.024506069339879328 | MAE: 0.0016519329651448987 | MAPE: 0.16697418179831902


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 110436653


Epoch: 150 | Inputs: 10 | hidden_size: 120 | Loss: MQLoss() | Scaler Type: standard | factor: 4 | decoder_layers: 1
Learning Rate: 2 | decoder_input_size_multiplier: 0.7 | Dropout: 0.3 | n_head: 2 | MovingAvg_window: 15| conv_hidden_size:32
RMSE: 0.0027057069595073976 | QLIKE: 0.026017404876643526 | MAE: 0.0016766413583528246 | MAPE: 0.16820847143281878


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 550 | Inputs: 42 | hidden_size: 150 | Loss: MSE() | Scaler Type: standard | factor: 4 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.3 | Dropout: 0 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.0029140317241152826 | QLIKE: 0.0314577865817137 | MAE: 0.0018830157680246747 | MAPE: 0.1871712398133047


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  if n_inputs[i]==5:
    a2=0
  if n_inputs[i]==15:
    a2=0
  else:
    a2=a2
  model = Autoformer(h=1, input_size=n_inputs[i],
             decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
             decoder_layers=decoder_layers[a5],
             MovingAvg_window=15, loss=losses[m],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | MovingAvg_window: {15}| conv_hidden_size:{conv_hidden_size[a3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 10472761


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 112800535


Epoch: 750 | Inputs: 84 | hidden_size: 180 | Loss: MSE() | Scaler Type: robust | factor: 4 | decoder_layers: 2
Learning Rate: 1 | decoder_input_size_multiplier: 0.4 | Dropout: 0 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:52
RMSE: 0.003049379710216954 | QLIKE: 0.03370670859904974 | MAE: 0.0020097719302826856 | MAPE: 0.203503283335252


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 91608222


Epoch: 250 | Inputs: 21 | hidden_size: 60 | Loss: MSE() | Scaler Type: robust | factor: 3 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.6 | Dropout: 0 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:12
RMSE: 0.0031523841900671285 | QLIKE: 0.03523803317346282 | MAE: 0.0020995708100581593 | MAPE: 0.212676004803766


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 59807235


Epoch: 650 | Inputs: 84 | hidden_size: 150 | Loss: MSE() | Scaler Type: robust | factor: 4 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.4 | Dropout: 0 | n_head: 3 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.0030318707305624906 | QLIKE: 0.033047394928424526 | MAE: 0.002039019514250593 | MAPE: 0.20995204493149808


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 56290688


Epoch: 350 | Inputs: 15 | hidden_size: 180 | Loss: DistributionLoss() | Scaler Type: robust | factor: 2 | decoder_layers: 2
Learning Rate: 2 | decoder_input_size_multiplier: 0.4 | Dropout: 0.4 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.002720475150864572 | QLIKE: 0.02660223651701244 | MAE: 0.001681759811369852 | MAPE: 0.1637382859319084


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 100 | Inputs: 42 | hidden_size: 180 | Loss: MQLoss() | Scaler Type: standard | factor: 2 | decoder_layers: 1
Learning Rate: 2 | decoder_input_size_multiplier: 0.3 | Dropout: 0.3 | n_head: 3 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.0037715752585198744 | QLIKE: 0.05346520837738813 | MAE: 0.002563653484989726 | MAPE: 0.25989336585195316


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  if n_inputs[i]==5:
    a2=0
  if n_inputs[i]==15:
    a2=0
  else:
    a2=a2
  model = Autoformer(h=1, input_size=n_inputs[i],
             decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
             decoder_layers=decoder_layers[a5],
             MovingAvg_window=15, loss=losses[m],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | MovingAvg_window: {15}| conv_hidden_size:{conv_hidden_size[a3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 14076085


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 3926335


Epoch: 450 | Inputs: 84 | hidden_size: 180 | Loss: MQLoss() | Scaler Type: standard | factor: 3 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.5 | Dropout: 0.3 | n_head: 3 | MovingAvg_window: 15| conv_hidden_size:32
RMSE: 0.0034706394256929234 | QLIKE: 0.044822514811979776 | MAE: 0.002275080987875502 | MAPE: 0.22361294564146092


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 26364829


Epoch: 100 | Inputs: 5 | hidden_size: 120 | Loss: MSE() | Scaler Type: standard | factor: 2 | decoder_layers: 3
Learning Rate: 0 | decoder_input_size_multiplier: 0.3 | Dropout: 0 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:42
RMSE: 0.0026491028539184064 | QLIKE: 0.02404413313384625 | MAE: 0.0016266785621670342 | MAPE: 0.16459221832717721


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 63182242


Epoch: 50 | Inputs: 10 | hidden_size: 90 | Loss: MAE() | Scaler Type: standard | factor: 2 | decoder_layers: 3
Learning Rate: 1 | decoder_input_size_multiplier: 0.7 | Dropout: 0.2 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:52
RMSE: 0.0028157464713639176 | QLIKE: 0.02921804926610903 | MAE: 0.00175037936558785 | MAPE: 0.17218788467182977


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 39293375


Epoch: 150 | Inputs: 5 | hidden_size: 90 | Loss: MQLoss() | Scaler Type: minmax | factor: 2 | decoder_layers: 2
Learning Rate: 2 | decoder_input_size_multiplier: 0.4 | Dropout: 0.3 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.002805381929791188 | QLIKE: 0.028435739123935072 | MAE: 0.0017143363801914474 | MAPE: 0.16810990552035687


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 250 | Inputs: 42 | hidden_size: 120 | Loss: MQLoss() | Scaler Type: standard | factor: 3 | decoder_layers: 3
Learning Rate: 0 | decoder_input_size_multiplier: 0.6 | Dropout: 0.3 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.0030393518990663886 | QLIKE: 0.034274578978738694 | MAE: 0.0019238092854621512 | MAPE: 0.1832730857297313


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  if n_inputs[i]==5:
    a2=0
  if n_inputs[i]==15:
    a2=0
  else:
    a2=a2
  model = Autoformer(h=1, input_size=n_inputs[i],
             decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
             decoder_layers=decoder_layers[a5],
             MovingAvg_window=15, loss=losses[m],
             max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
             val_check_steps=500, batch_size=32,
             scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | MovingAvg_window: {15}| conv_hidden_size:{conv_hidden_size[a3]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 76707156


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 127675984


Epoch: 150 | Inputs: 15 | hidden_size: 150 | Loss: MSE() | Scaler Type: robust | factor: 2 | decoder_layers: 1
Learning Rate: 2 | decoder_input_size_multiplier: 0.5 | Dropout: 0 | n_head: 3 | MovingAvg_window: 15| conv_hidden_size:52
RMSE: 0.0027768420355346315 | QLIKE: 0.02753153966242932 | MAE: 0.0017647280960971451 | MAPE: 0.17788091267377923


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 107737853


Epoch: 100 | Inputs: 5 | hidden_size: 180 | Loss: MSE() | Scaler Type: minmax | factor: 2 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.4 | Dropout: 0 | n_head: 4 | MovingAvg_window: 15| conv_hidden_size:42
RMSE: 0.002711240059899987 | QLIKE: 0.02516893188248694 | MAE: 0.0016667021040866782 | MAPE: 0.16827030117283595


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 71096575


Epoch: 650 | Inputs: 15 | hidden_size: 90 | Loss: MSE() | Scaler Type: minmax | factor: 2 | decoder_layers: 3
Learning Rate: 0 | decoder_input_size_multiplier: 0.7 | Dropout: 0 | n_head: 3 | MovingAvg_window: 15| conv_hidden_size:42
RMSE: 0.0029753179491293787 | QLIKE: 0.033045273449646494 | MAE: 0.0019501443616643044 | MAPE: 0.2005617605013934


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 41135338


Epoch: 650 | Inputs: 21 | hidden_size: 150 | Loss: MSE() | Scaler Type: robust | factor: 2 | decoder_layers: 3
Learning Rate: 1 | decoder_input_size_multiplier: 0.5 | Dropout: 0 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:52
RMSE: 0.0026905831663233386 | QLIKE: 0.02566067061187306 | MAE: 0.0016874540358190967 | MAPE: 0.16878288236706826


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 750 | Inputs: 5 | hidden_size: 150 | Loss: DistributionLoss() | Scaler Type: minmax | factor: 2 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.5 | Dropout: 0.4 | n_head: 2 | MovingAvg_window: 15| conv_hidden_size:22
RMSE: 0.0030188674592545096 | QLIKE: 0.03625652190067799 | MAE: 0.0018455521369970828 | MAPE: 0.18569634350330846


INFO:lightning_fabric.utilities.seed:Seed set to 289990

Epoch: 750 | Inputs: 5 | hidden_size: 120 | Loss: MAE() | Scaler Type: robust | factor: 2 | decoder_layers: 2
Learning Rate: 0 | decoder_input_size_multiplier: 0.4 | Dropout: 0.2 | n_head: 6 | MovingAvg_window: 15| conv_hidden_size:32
RMSE: 0.0026014803693844026 | QLIKE: 0.023928809728980475 | MAE: 0.0015433653548939167 | MAPE: 0.1489220579925518

In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = Autoformer(h=1, input_size=5,
             decoder_input_size_multiplier=0.4, hidden_size=120,
           dropout=0.2, factor=2, n_head=6,
           conv_hidden_size=32,
             decoder_layers=2,
             MovingAvg_window=15, loss=losses[1],
             max_steps=750,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
             val_check_steps=500, batch_size=32,
             scaler_type='robust',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 70936270


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 26591285


Random Seed: 70936270
RMSE: 0.002615960975370439 | QLIKE: 0.024127321319439263 | MAE: 0.0015565249042460193 | MAPE: 0.15003370453078163


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 81284161


Random Seed: 26591285
RMSE: 0.002617821129571827 | QLIKE: 0.02417452636003432 | MAE: 0.0015574228413888524 | MAPE: 0.14985807775448018


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 100849848


Random Seed: 81284161
RMSE: 0.0026026728700798613 | QLIKE: 0.023911986003791706 | MAE: 0.0015452382778488247 | MAPE: 0.14876282529169735


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 64113249


Random Seed: 100849848
RMSE: 0.002626973198101482 | QLIKE: 0.02441024290031015 | MAE: 0.0015636675597552813 | MAPE: 0.15005484476171174


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 64113249
RMSE: 0.0026159619758729545 | QLIKE: 0.024138035876253617 | MAE: 0.0015518282598773682 | MAPE: 0.14933247211726608


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = Autoformer(h=1, input_size=5,
             decoder_input_size_multiplier=0.4, hidden_size=120,
           dropout=0.2, factor=2, n_head=6,
           conv_hidden_size=32,
             decoder_layers=2,
             MovingAvg_window=15, loss=losses[1],
             max_steps=750,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
             val_check_steps=500, batch_size=32,
             scaler_type='robust',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 123432993


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 65233172


Random Seed: 123432993
RMSE: 0.002623516718228541 | QLIKE: 0.024271005988813557 | MAE: 0.001558447497686376 | MAPE: 0.15016169253919537


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 38983352


Random Seed: 65233172
RMSE: 0.002626569534054346 | QLIKE: 0.02435250132748366 | MAE: 0.0015609306487048256 | MAPE: 0.14988132248042407


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 77739580


Random Seed: 38983352
RMSE: 0.002626799470946596 | QLIKE: 0.02432593837419286 | MAE: 0.0015682105072290981 | MAPE: 0.1509432985695467


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 80292897


Random Seed: 77739580
RMSE: 0.002608834402092804 | QLIKE: 0.02411353913494415 | MAE: 0.0015518055923746347 | MAPE: 0.14969723542665883


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 80292897
RMSE: 0.002638165546353872 | QLIKE: 0.024427735670265495 | MAE: 0.0015743939379520462 | MAPE: 0.15183510045932794


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = Autoformer(h=1, input_size=5,
             decoder_input_size_multiplier=0.4, hidden_size=120,
           dropout=0.2, factor=2, n_head=6,
           conv_hidden_size=32,
             decoder_layers=2,
             MovingAvg_window=15, loss=losses[1],
             max_steps=750,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
             val_check_steps=500, batch_size=32,
             scaler_type='robust',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 18066954


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 80970279


Random Seed: 18066954
RMSE: 0.002616249715483087 | QLIKE: 0.024137419082759265 | MAE: 0.0015563941369749921 | MAPE: 0.14987215960667666


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 111831793


Random Seed: 80970279
RMSE: 0.0026219116706046766 | QLIKE: 0.024193361245873506 | MAE: 0.0015718603756406434 | MAPE: 0.15191721710154169


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 25592992


Random Seed: 111831793
RMSE: 0.0026183583635138396 | QLIKE: 0.024228289402597795 | MAE: 0.0015573301194398148 | MAPE: 0.14962141984308377


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 45175497


Random Seed: 25592992
RMSE: 0.002620520833757006 | QLIKE: 0.024180845338187866 | MAE: 0.001562214187007338 | MAPE: 0.15014924727502632


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 45175497
RMSE: 0.00262837988890831 | QLIKE: 0.024434129430527057 | MAE: 0.0015758797359756859 | MAPE: 0.15187126054586858


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = Autoformer(h=1, input_size=5,
             decoder_input_size_multiplier=0.4, hidden_size=120,
           dropout=0.2, factor=2, n_head=6,
           conv_hidden_size=32,
             decoder_layers=2,
             MovingAvg_window=15, loss=losses[1],
             max_steps=750,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
             val_check_steps=500, batch_size=32,
             scaler_type='robust',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Autoformer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Autoformer"].values
  else:
    Y_hat=forecasts["Autoformer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 14276172


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 52712633


Random Seed: 14276172
RMSE: 0.0026228710283354973 | QLIKE: 0.024219094028143658 | MAE: 0.0015628961530032382 | MAPE: 0.1504191768922914


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 34256885


Random Seed: 52712633
RMSE: 0.0026269626830777108 | QLIKE: 0.024341762075369245 | MAE: 0.0015738920914581465 | MAPE: 0.15184321947374613


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 109128094


Random Seed: 34256885
RMSE: 0.0026296033000236256 | QLIKE: 0.024396707206520235 | MAE: 0.001571763386523049 | MAPE: 0.15157857327816587


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 115818663


Random Seed: 109128094
RMSE: 0.0026319882103398784 | QLIKE: 0.024438042256080417 | MAE: 0.0015717623318392527 | MAPE: 0.1510232325086402


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 115818663
RMSE: 0.0026346421327691125 | QLIKE: 0.024392042685290966 | MAE: 0.0015719645391619253 | MAPE: 0.1511529141088067


# Test

In [6]:
test_length=int(len(df['NKE'])*0.3)
df1=df.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

# H=1

In [ ]:
model = Autoformer(h=1, input_size=5,
             decoder_input_size_multiplier=0.4, hidden_size=120,
           dropout=0.2, factor=2, n_head=6,
           conv_hidden_size=32,
             decoder_layers=2,
             MovingAvg_window=15, loss=losses[1],
             max_steps=750,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
             val_check_steps=500, batch_size=32,
             scaler_type='robust',
              random_seed=289990)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=1,static_df=df_static,n_windows=None, test_size=test_length-test_length%1,step_size=1)
forecasts = forecasts.dropna()
if "Autoformer-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["Autoformer"].values
else:
  Y_hat=forecasts["Autoformer-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 289990


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.0038742636604709394 | QLIKE: 0.03130798284496256 | MAE: 0.0023975643290157444 | MAPE: 0.1796906327216747


In [ ]:
if "Autoformer-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","Autoformer-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

In [ ]:
df2.to_excel("Autoformer forecasts.xlsx")

# H=5

In [7]:
H=5

model = Autoformer(h=H, input_size=5,
             decoder_input_size_multiplier=0.4, hidden_size=120,
           dropout=0.2, factor=2, n_head=6,
           conv_hidden_size=32,
             decoder_layers=2,
             MovingAvg_window=15, loss=losses[1],
             max_steps=750,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
             val_check_steps=500, batch_size=32,
             scaler_type='robust',
              random_seed=289990)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=H,static_df=df_static,n_windows=None, test_size=test_length-test_length%H,step_size=H)
forecasts = forecasts.dropna()
if "Autoformer-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["Autoformer"].values
else:
  Y_hat=forecasts["Autoformer-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

if "Autoformer-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","Autoformer-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

df2.to_excel("Autoformer forecasts.xlsx")

INFO:lightning_fabric.utilities.seed:Seed set to 289990


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.004709075667739795 | QLIKE: 0.04569864820047234 | MAE: 0.0028575014925553166 | MAPE: 0.21530123687944494


# H=10

In [8]:
H=10

model = Autoformer(h=H, input_size=5,
             decoder_input_size_multiplier=0.4, hidden_size=120,
           dropout=0.2, factor=2, n_head=6,
           conv_hidden_size=32,
             decoder_layers=2,
             MovingAvg_window=15, loss=losses[1],
             max_steps=750,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
             val_check_steps=500, batch_size=32,
             scaler_type='robust',
              random_seed=289990)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=H,static_df=df_static,n_windows=None, test_size=test_length-test_length%H,step_size=H)
forecasts = forecasts.dropna()
if "Autoformer-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["Autoformer"].values
else:
  Y_hat=forecasts["Autoformer-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

if "Autoformer-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","Autoformer-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

df2.to_excel("Autoformer forecasts.xlsx")

INFO:lightning_fabric.utilities.seed:Seed set to 289990


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.00570747087072068 | QLIKE: 0.060183553127669286 | MAE: 0.0033522360643235133 | MAPE: 0.24450347549933338
